### PR TITLE
E2E storage: more tests for different pod/node combinations

### PR DESCRIPTION
**Pull request by pohly**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

There was no test for running multiple pods at the same time one the same node, sharing the same volume. Such a test is relevant for CSI drivers which need to implement `NodeStageVolume` for this kind of usage to work.

Also not covered was accessing data on one node after writing it on another.

**Special notes for your reviewer**:

In order to implement the new test cleanly, the existing test infrastructure and other tests get refactored first.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

/sig storage
/cc @msau42

### CODE REVIEW

**Comment by pohly** [comment]
I am not sure whether it makes sense to enable this test for all test patterns. A full matrix (all tests for all configurations) of course reduces the chance of missing some situation where the test fails, but it also increases the overall test runtime.

For now I am following the "testsuites" approach where everything is enabled for all test patterns.

**Comment by pohly** [comment]
/hold 

Waiting for other PRs to get merged first.

**Comment by pohly** [comment]
@msau42 I ended up using the approach where the provisioning test determines which nodes it has first, and then picks nodes by name. I think it is simpler this way and it automatically handles the case where the node selector only matches a single node.

**Comment by pohly** [comment]
/hold cancel

It doesn't look like my other, conflicting PRs are going to be merged soon. Therefore I have rebased this PR onto master, i.e. it could be merged now.

@msau42 please take a look

**Comment by pohly** [comment]
@msau42 the gcepd driver fails the new "use PVC in two pods on the same node" test":
```
I0109 08:29:11.168] Jan  9 08:29:03.048: INFO: At 2019-01-09 08:13:14 +0000 UTC - event for pvc-volume-tester-second: {attachdetach-controller } FailedAttachVolume: Multi-Attach error for volume "pvc-63518474-13e6-11e9-8c13-42010a280002" Volume is already used by pod(s) pvc-volume-tester-first
```

Shall I disable the test by marking the `gcepd` with `Capabilities[CapMultiPODs] = false`, or are so many drivers going to fail this test that the default should be "disabled unless explicitly enabled"?

**Comment by pohly** [comment]
I suppose it comes down to this CSI PR: https://github.com/container-storage-interface/spec/pull/150

According to the updated definition of `SINGLE_WRITER`, a volume must not be used by more than one pod on a node, even though that would be technically feasible via bind-mounting. That makes the test invalid, because it relies on undefined behavior.

Whether the driver needs to detect the condition is a different matter.

@msau42 agreed?

I wonder under which circumstances `NodeStageVolume` then is needed. Perhaps rapidly starting different pods on the same node using the same volume (i.e. one `NodeStageVolume`, followed by multiple `NodePublishVolume/NodeUnpublishVolume` pairs)?

**Comment by pohly** [comment]
The test would still make sense for `MULTI_SINGLE_WRITER` and higher, which can be detected at test execution time. Defining a test `Capability` is just a shortcut.

File: test/e2e/storage/testsuites/testdriver.go [hunk 1]
```diff
@@ -78,14 +78,17 @@ type DynamicPVTestDriver interface {
...
+// Unless noted otherwise, capabilities are assumed to be not
+// supported when not set explicitly.
 type Capability string
 
 const (
 	CapPersistence Capability = "persistence" // data is persisted across pod restarts
 	CapBlock       Capability = "block"       // raw block mode
 	CapFsGroup     Capability = "fsGroup"     // volume ownership via fsGroup
 	CapExec        Capability = "exec"        // exec a file in the volume
+	CapMultiPODs   Capability = "multipods"   // multiple pods on a node can use the same volume concurrently; enabled by default
```
Comment (wongma7) [inline test/e2e/storage/testsuites/testdriver.go]: how/where is this enabled by default?

Comment (pohly) [reply test/e2e/storage/testsuites/testdriver.go]: At the place where it is getting checked. If unset, the code assumes that the driver supports multiple pods:
https://github.com/kubernetes/kubernetes/blob/5551371f0dde5ec7be93659d2cce4d92fa3f6eca/test/e2e/storage/testsuites/provisioning.go#L201-L205

But I am no longer sure what the correct behavior is in this situation. Technically it can work and I thought `NodeStageVolume` was meant to enable this double use on the same node, but the CSI spec PR makes it clear that this usage was not meant to be supported.

Comment (wongma7) [reply test/e2e/storage/testsuites/testdriver.go]: I don't think the spec being insufficient means this test is invalid or that this test is an endorsement of unsupported behaviour. My understanding is that this behaviour (ability for multiple pods on 1 node to write to the same volume) is not explicitly called out in kubernetes docs but it follows naturally from the way kubernetes does mounts and defines access modes (RWO) in terms of nodes, not pods. So many users rely on it.

We know that the GCE PD CSI driver supports this capability even though it reports SINGLE_WRITER, we know SINGLE_WRITER isn't enough to capture the driver's capabilities...rather than wait for the spec to update to accurately capture what kubernetes has allowed plugins/drivers/users to do since before CSI, we can add the test now and set the capability true for drivers as we see fit (this means disabling this bool by default like the others, which we should do anyway IMO for consistency).

Seems the latest discussion on the spec issue is here: https://github.com/container-storage-interface/spec/issues/178.

File: test/e2e/storage/testsuites/provisioning.go [hunk 2]
```diff
@@ -195,6 +197,16 @@ func testProvisioning(input *provisioningTestInput) {
 		input.pvc.Spec.VolumeMode = &block
 		TestDynamicProvisioning(input.testCase, input.cs, input.pvc, input.sc)
 	})
+
+	It("should allow concurrent writes on the same node", func() {
```
Comment (wongma7) [inline test/e2e/storage/testsuites/provisioning.go]: I don't think the test belongs in the provisioning suite. volumes might be a better fit? I guess it will need more boilerplate though since we can't piggyback on TestDynamicProvisioning.

Comment (wongma7) [reply test/e2e/storage/testsuites/provisioning.go]: it might be okay if we make CapMultiPODs non-default though, then I would have to explicitly sign up my driver for this test anyway. Up to you..

File: test/e2e/storage/testsuites/testdriver.go (hunk 1 above)
Comment (pohly) [reply test/e2e/storage/testsuites/testdriver.go]: I agree that this spec limitation seems rather artificial and I am fine with having a test that explicitly tests support for this particular even if it is not covered by the spec (due to #178 being open).

File: test/e2e/storage/testsuites/provisioning.go (hunk 2 above)
Comment (pohly) [reply test/e2e/storage/testsuites/provisioning.go]: Yes, let's do that. My original reasoning was that all drivers must pass the test and therefore nothing special should be needed to run it, but as that's not the case, let's make it opt-in.

Comment (pohly) [reply test/e2e/storage/testsuites/provisioning.go]: Regarding where to put the test: I agree that this might belong into its own test suite, but it is not clear to me which one ("volume-features"?) and now is a bad time to introduce one because of the concurrent work on PR #72434. Let's keep it here, okay?

**Comment by pohly** [comment]
Test is now enabled only for CSI hostpath. I want to start small because I am not sure anymore which drivers are expected to pass and which jobs cover them.

Bug for gecpd is here: https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/issues/219

File: test/e2e/storage/testsuites/provisioning.go (hunk 2 above)
Comment (wongma7) [reply test/e2e/storage/testsuites/provisioning.go]: agree

**Comment by pohly** [comment]
@wongma7 can you have another look at this PR?

File: test/e2e/storage/testsuites/provisioning.go [hunk 3]
```diff
@@ -283,11 +313,54 @@ func TestDynamicProvisioning(t StorageClassTest, client clientset.Interface, cla
...
+		wg := sync.WaitGroup{}
+		wg.Add(2)
+		run := func(suffix, command string) {
+			defer GinkgoRecover()
+			defer wg.Done()
+			runInPodWithVolume(client, claim.Namespace, claim.Name, suffix, nodeName, command, nil /* node selector */, t.ExpectUnschedulable)
+		}
+		go run("-first", "touch /mnt/test/first && while ! [ -f /mnt/test/second ]; do sleep 1; done")
+		go run("-second", "touch /mnt/test/second && while ! [ -f /mnt/test/first ]; do sleep 1; done")
+		wg.Wait()
```
Comment (wongma7) [inline test/e2e/storage/testsuites/provisioning.go]: should we have a timeout here? how long is the default timeout for a long test (if there is one)?

File: test/e2e/storage/testsuites/provisioning.go [hunk 4]
```diff
@@ -396,16 +469,17 @@ func TestBindingWaitForFirstConsumerMultiPVC(t StorageClassTest, client clientse
...
 // runInPodWithVolume runs a command in a pod with given claim mounted to /mnt directory.
-func runInPodWithVolume(c clientset.Interface, ns, claimName, nodeName, command string, nodeSelector map[string]string, unschedulable bool) {
+func runInPodWithVolume(c clientset.Interface, ns, claimName, suffix, nodeName, command string, nodeSelector map[string]string, unschedulable bool) {
 	pod := &v1.Pod{
 		TypeMeta: metav1.TypeMeta{
 			Kind:       "Pod",
 			APIVersion: "v1",
 		},
 		ObjectMeta: metav1.ObjectMeta{
-			GenerateName: "pvc-volume-tester-",
```
Comment (wongma7) [inline test/e2e/storage/testsuites/provisioning.go]: I think we should keep using GenerateName, like 

GenerateName: "pvc-volume-tester-" + suffix,

Will make it easier to find something in a log

Comment (pohly) [reply test/e2e/storage/testsuites/provisioning.go]: Because of the random number at the end? Okay, I see how that may help. I'll add it back.

File: test/e2e/storage/testsuites/provisioning.go (hunk 3 above)
Comment (pohly) [reply test/e2e/storage/testsuites/provisioning.go]: `runInPodWithVolume` has its own timeout, 5 minutes I think (from ` framework.WaitForPodSuccessInNamespaceSlow`) - basically the same as in all the other tests. I don't think we need another one.

File: test/e2e/storage/testsuites/provisioning.go [hunk 5]
```diff
@@ -283,11 +313,54 @@ func TestDynamicProvisioning(t StorageClassTest, client clientset.Interface, cla
...
+
+		By("checking the created volume is writable")
+		runInPodWithVolume(client, claim.Namespace, claim.Name, "-first", nodeName, command,
+			nil /* node selector */, false /* expect unschedulable */)
 
 		By("checking the created volume is readable and retains data")
-		runInPodWithVolume(client, claim.Namespace, claim.Name, t.NodeName, "grep 'hello world' /mnt/test/data", t.NodeSelector, t.ExpectUnschedulable)
+		runInPodWithVolume(client, claim.Namespace, claim.Name, "-second", nodeName, "grep 'hello world' /mnt/test/data", nil /* node selector */, t.ExpectUnschedulable)
+
+		// Run on another node, if we have one.
```
Comment (msau42) [inline test/e2e/storage/testsuites/provisioning.go]: Picking the 2nd node should still respect any node selectors set.  For example, for GCE PD, the 2nd node needs to be in the same zone as the 1st node.

File: test/e2e/storage/testsuites/provisioning.go [hunk 6]
```diff
@@ -396,16 +469,17 @@ func TestBindingWaitForFirstConsumerMultiPVC(t StorageClassTest, client clientse
...
 		TypeMeta: metav1.TypeMeta{
 			Kind:       "Pod",
 			APIVersion: "v1",
 		},
 		ObjectMeta: metav1.ObjectMeta{
-			GenerateName: "pvc-volume-tester-",
+			Name: "pvc-volume-tester" + suffix,
 		},
 		Spec: v1.PodSpec{
+			NodeName: nodeName,
```
Comment (msau42) [inline test/e2e/storage/testsuites/provisioning.go]: Can we specify the node name as a node selector instead?  Setting NodeName directly in the pod spec will completely bypass the scheduler, and we will miss testing our volume-related scheduling logic.

**Comment by msau42** [comment]
/assign @verult 
to look into any potential impact to topology and repd tests with the changes to `TestDynamicProvisioning`

File: test/e2e/storage/testsuites/provisioning.go [hunk 7]
```diff
@@ -198,6 +199,15 @@ func testProvisioning(input *provisioningTestInput) {
...
 		TestDynamicProvisioning(input.testCase, input.cs, input.pvc, input.sc)
 	})
+
+	It("should allow concurrent writes on the same node", func() {
+		if !input.dInfo.Capabilities[CapMultiPODs] {
+			framework.Skipf("Driver %q does not support multiple concurrent pods - skipping", input.dInfo.Name)
+		}
+		input.testCase.SkipWriteReadCheck = true
+		input.testCase.MultiWriteCheck = true
+		TestDynamicProvisioning(input.testCase, input.cs, input.pvc, input.sc)
```
Comment (msau42) [inline test/e2e/storage/testsuites/provisioning.go]: I know this function does a lot of common setup, but I think it's going to get harder to understand/read as we add more functionality that only pertains to one test case.

Instead, can we move the setup into a common function, and keep each test-case specific logic in its own function?

File: test/e2e/storage/testsuites/provisioning.go (hunk 5 above)
Comment (pohly) [reply test/e2e/storage/testsuites/provisioning.go]: The two test pods run on the exact same node, and that node is guaranteed to come from the nodes selected for the test via the node selector. Isn't that what you are asking for?

File: test/e2e/storage/testsuites/provisioning.go (hunk 6 above)
Comment (pohly) [reply test/e2e/storage/testsuites/provisioning.go]: Yes, that's possible with the `kubernetes.io/hostname` label. I'll change it.

File: test/e2e/storage/testsuites/provisioning.go (hunk 7 above)
Comment (pohly) [reply test/e2e/storage/testsuites/provisioning.go]: I'll have a look at that.

**Review by pohly — COMMENTED** [review]
I'll add some more log output that shows where pods land and then enable the gcepd driver again. That should show whether the two concurrent pods really run on the same node.

File: test/e2e/storage/testsuites/provisioning.go (hunk 7 above)
Comment (pohly) [reply test/e2e/storage/testsuites/provisioning.go]: Why do I always end up cleaning up tests written by others?! :tired_face:

`StorageClassTest.ExpectUnschedulable` was added very recently and now I need to move that logic to the one test case that actually uses it. Just saying.

File: test/e2e/storage/testsuites/provisioning.go (hunk 5 above)
Comment (pohly) [reply test/e2e/storage/testsuites/provisioning.go]: Perhaps I need to point out that TestDynamicProvisioning never (?!) ran with a non-zero `t.NodeSelector`. `t.NodeSelector` itself was introduced fairly recently (https://github.com/kubernetes/kubernetes/commit/12ece9dacbbe8b3f9727c5136912de9089b54b27#diff-6dc833c65b9c544adfb2d654d2f7cbd3) and even after that commit, `testProvisioning` did not set it (https://github.com/kubernetes/kubernetes/blob/12ece9dacbbe8b3f9727c5136912de9089b54b27/test/e2e/storage/testsuites/provisioning.go#L87-L100).

As far as I can tell, `t.NodeSelector` was introduced for the `TestBinding` tests and is only used there.

In other words, my changes shouldn't cause the test to stop using a node selector.

File: test/e2e/storage/testsuites/provisioning.go (hunk 4 above)
Comment (pohly) [reply test/e2e/storage/testsuites/provisioning.go]: Done.

File: test/e2e/storage/testsuites/provisioning.go (hunk 7 above)
Comment (pohly) [reply test/e2e/storage/testsuites/provisioning.go]: I ended up not touching `StorageClassTest.ExpectUnschedulable`. IMHO it would be better to have that as a separate parameter for those functions which actually use it, but that would have led to even more code churn. It was false in all cases for `TestDynamicProvisioning` and that function now ignores it. Having it at `true` wouldn't make sense, because why run that function in a configuration where the two test pods cannot be scheduled?

What I have done is move the write/read check out of `TestDynamicProvisioning`. Callers of `TestDynamicProvisioning` now decide which additional checks they want to run on the PV, which makes it a bit easier to understand what each testcase really covers. The "run two pods concurrently" check then hooks into `TestDynamicProvisioning` the same way.

The first three commits in this PR are now generic cleanup commits. If you want, I can submit them separately.

File: test/e2e/storage/testsuites/provisioning.go [hunk 8]
```diff
@@ -306,6 +316,95 @@ func TestDynamicProvisioning(t StorageClassTest, client clientset.Interface, cla
...
+// - the second node is only used if nodeSelector is set and matches more than
+//   the first node
+// - if both nodeName and nodeSelector are unset, then the test cannot lock
+//   the pod onto the same node
+//
+// If all succeeds, Kubernetes actually allocated something that is
+// persistent across pods and (optional) across nodes.
+//
+// This is a common test that can be called from a StorageClassTest.PvCheck.
+func PVWriteReadCheck(client clientset.Interface, claim *v1.PersistentVolumeClaim, volume *v1.PersistentVolume, nodeName string, nodeSelector map[string]string) {
```
Comment (msau42) [inline test/e2e/storage/testsuites/provisioning.go]: Instead of modifying the behavior of the existing check, can we add a new check that forces scheduling on a different node?

My thought is that this should be an explicit new test case.  And that the test case should not have different behavior depending on the driver, ie for hostpath driver, this test case is actually not going to test what it wants to test since hostpath only works on a single node.

File: test/e2e/storage/testsuites/provisioning.go [hunk 9]
```diff
@@ -306,6 +316,95 @@ func TestDynamicProvisioning(t StorageClassTest, client clientset.Interface, cla
...
+	} else {
+		By("checking the created volume is readable and retains data on some node")
+	}
+	command = "grep 'hello world' /mnt/test/data"
+	runInPodWithVolume(client, claim.Namespace, claim.Name, "-second", "" /* node name */, command, firstNodeSelector, false)
+
+	// Run on another node, if we have one.
+	var secondNodeName string
+	if nodes != nil {
+		for _, node := range nodes.Items {
```
Comment (msau42) [inline test/e2e/storage/testsuites/provisioning.go]: In a multizone cluster nodes are going to be in multiple zones.  Blindly picking any node could conflict with PV.nodeaffinity.

I would suggest instead setting anti affinity to the first node as a way to pick the second node.

File: test/e2e/storage/testsuites/provisioning.go (hunk 7 above)
Comment (msau42) [reply test/e2e/storage/testsuites/provisioning.go]: Thanks for finding a way to do this. I wasn't expecting you to clean up other tests.  I was thinking to wrap setup in a function, change TestDynamicProvisioning to use the function, and then write a new function for your test case.  But this works too.

File: test/e2e/storage/testsuites/provisioning.go (hunk 8 above)
Comment (pohly) [reply test/e2e/storage/testsuites/provisioning.go]: So in other words, have one test for single node and another test for multi-node?

What I don't like about the current test is that it's non-deterministic on a multi-node cluster. The second pod may or may not end up on the same node.

File: test/e2e/storage/testsuites/provisioning.go (hunk 9 above)
Comment (pohly) [reply test/e2e/storage/testsuites/provisioning.go]: The intended usage for this test is so that the node selector matches all nodes that are compatible with the PV. I guess need to spell this out more explicitly.

When used like that, the problem you mentioned cannot happen, can it?

File: test/e2e/storage/testsuites/provisioning.go (hunk 8 above)
Comment (msau42) [reply test/e2e/storage/testsuites/provisioning.go]: Yes, I think we should have explicit test cases for the two scenarios.

The current test is non-deterministic, however, it is more realistic to what a user would experience.

**Comment by pohly** [comment]
Michelle Au <notifications@github.com> writes:
> Thanks for finding a way to do this. I wasn't expecting you to clean
> up other tests.

We can't keep adding new tests by adding new fields, parameters and if
checks. At some point we have to take the time to clean up.

> I was thinking to wrap setup in a function, change
> TestDynamicProvisioning to use the function,

I had considered that, but the preamble of TestDynamicProvisioning
depends on `defer` in a number of places, which doesn't work well when
that preamble gets moved to a separate setup function.

> and then write a new
> function for your test case.  But this works too.

Good, then let's continue like this.

File: test/e2e/storage/testsuites/provisioning.go (hunk 9 above)
Comment (msau42) [reply test/e2e/storage/testsuites/provisioning.go]: The node selectors were originally a workaround when we didn't have a concept of topology for PVs and we needed to restrict the pods to specific zones.  Now that we have PV.NodeAffinity, we shouldn't need to restrict pods with a node selector.

Comment (msau42) [reply test/e2e/storage/testsuites/provisioning.go]: As a side note, topology in CSI is still alpha and disabled by default.  For now, we have disabled multi-zone CSI GCE PD tests until topology goes beta.  So I think we should still design the test case around leveraging PV topology.

File: test/e2e/storage/testsuites/provisioning.go (hunk 8 above)
Comment (pohly) [reply test/e2e/storage/testsuites/provisioning.go]: I'm not convinced that multiple test cases is really desirable. If we really wanted to focus each test on a single aspect, we would need three tests
- "can provision a PV"
- "can write/read on single node"
- "can write/read on different nodes"

If already provisioning fails, there's no point in testing the other two cases. If the (presumably simpler) write/read on a single node fails, that should be investigated first. But I see that this is a bit subjective.

BTW, the write/read test was extended to also test mount flags. Should that have been a separate test, too?

Anyway, I have split it up into two separate cases. Scheduling is based entirely on node and pod affinity now, at least for the first pod.

@msau42 please have another look

File: test/e2e/storage/testsuites/provisioning.go (hunk 9 above)
Comment (pohly) [reply test/e2e/storage/testsuites/provisioning.go]: The tests no longer pick a node themselves. Instead I let Kubernetes schedule a pod, then determine where it landed to schedule the second pod (by name for the "same node" test, by anti-affinity for the "other node"  test).

The provisioning test was only using the node name and ignored the node selector that was introduced for the topology tests. I've kept it that way.

**Comment by msau42** [comment]
Let's try this again
/test pull-kubernetes-e2e-gce-csi-serial

**Comment by msau42** [comment]
/approve

@verult can you help review the changes to the topology tests and run these on a multi-zone cluster?

**Comment by msau42** [comment]
Looks like the new tests passed (on a single zone cluster).  The reconstruction tests are expected to fail.

**Comment by pohly** [comment]
I'm currently waiting for these PRs to get merged first, then I'll have to rebase:
* https://github.com/kubernetes/kubernetes/pull/69036
* https://github.com/kubernetes/kubernetes/pull/70823

**Comment by pohly** [comment]
/hold

I have rebased on top of https://github.com/kubernetes/kubernetes/pull/69036 and now wait for that to be merged.

**Comment by pohly** [comment]
@msau42 I've included the fixes for https://github.com/kubernetes/kubernetes/pull/69036 and rebased.

Please have another look.

**Comment by pohly** [comment]
/hold

I need to check why `[sig-storage] CSI Volumes [Driver: csi-hostpath-v0] [Testpattern: Dynamic Snapshot] snapshottable should create snapshot with defaults [Feature:VolumeSnapshotDataSource]` isn't getting skipped.

**Comment by pohly** [comment]
Found it. Introducing the snapshot short-circuited some checks in `skipUnsupportedTests`. See f8536e8.

/hold cancel

**Comment by pohly** [comment]
@msau42 all tests are clean now in this PR.

File: test/e2e/storage/testsuites/provisioning.go [hunk 10]
```diff
@@ -340,15 +381,125 @@ func TestDynamicProvisioning(t StorageClassTest, client clientset.Interface, cla
...
+// - The second pod runs grep 'hello world' on /mnt/test on the same node.
+//
+// The node is selected by Kubernetes when scheduling the first
+// pod. It's then selected via its name for the second pod.
+//
+// If both succeed, Kubernetes actually allocated something that is
+// persistent across pods.
+//
+// This is a common test that can be called from a StorageClassTest.PvCheck.
+func PVWriteReadCheck(client clientset.Interface, claim *v1.PersistentVolumeClaim, volume *v1.PersistentVolume, node NodeSelection) {
```
Comment (msau42) [inline test/e2e/storage/testsuites/provisioning.go]: Can you rename this to PVWriteReadSingleNodeCheck to be more obvious about what the function is doing?

**Review by msau42 — COMMENTED** [review]
/approve

Just one nit from me. 

I was originally concerned that we would be losing test coverage by changing the RWCheck to only launch pods on a single node, but all the volume drivers tested in volume_provisioning.go should be covered in drivers/in_tree.go.  

I think it's also ok for the regional_pd.go tests that use RWCheck to only check a single node because we will get multi-node coverage through the failover test cases.

File: test/e2e/storage/testsuites/provisioning.go (hunk 10 above)
Comment (pohly) [reply test/e2e/storage/testsuites/provisioning.go]: Done.
